File: problem-archive/DP/L509_Fibonacci_Number.go
// Problem Source: https://leetcode.com/problems/fibonacci-number/
package main

import (
    "fmt"
    "os"
)

func fib(n int) int {
    if n < 1 { return 0 }
    cache := make([]int, n + 1)
    cache[0], cache[1] = 0, 1
    for i := 2; i <= n; i++ { cache[i] = cache[i - 1] + cache[i - 2] }
    return cache[n]
}

type Testing struct {
    n, expect   int
}

func main() {
    testCases := []*Testing{
        { n: 2, expect: 1 },
        { n: 3, expect: 2 },
        { n: 4, expect: 3 },
    }

    for _, t := range testCases {
        if outp := fib(t.n); outp != t.expect {
            fmt.Println("FAIL")
            fmt.Println("Want: ", t.expect)
            fmt.Println("Got: ", outp)
            os.Exit(1)
        }
    }

    fmt.Println("ok")
}

File: problem-archive/DP/L509_Fibonacci_Number.js
/**
 * @param {number} n
 * @return {number}
 */
var fib = function(n) {
    if (n < 1) {
        return 0;
    }

    let t0 = 0;
    let t1 = 1;
    let t2 = 0;
    for (let i = 1; i < n; i++) {
        t2 = t1;
        t1 += t0;
        t0 = t2;
    }

    return t1;
};

console.log(fib(2));
console.log(fib(3));
console.log(fib(4));
